integration_tests/waypoints.cpp

```diff
@@ -64,276 +37,171 @@ TEST_F(SitlTest, MissionAddWaypointsAndFly)
+                         10.0f, 5.0f, false,
+                         0.0f, 0.0f,
+                         MissionItem::CameraAction::STOP_PHOTO_INTERVAL));
+
+    {
+        // We only have the send_mission function asynchronous for now, so we wrap it using
+        // std::future.
+        auto prom = std::make_shared<std::promise<void>>();
+        auto future_result = prom->get_future();
+        device.mission().send_mission_async(
+        mission_items, [prom](Mission::Result result) {
+            EXPECT_EQ(result, Mission::Result::SUCCESS);
+            prom->set_value();
+        });
+
+        future_result.wait();
+        future_result.get();
     }
 
-    EXPECT_EQ(_mission_state, MissionState::DONE);
-}
+    const Action::Result arm_result = device.action().arm();
+    EXPECT_EQ(arm_result, Action::Result::SUCCESS);
 
-void receive_send_mission_result(Mission::Result result)
-{
-    EXPECT_EQ(result, Mission::Result::SUCCESS);
+    // Before starting the mission, we want to be sure to subscribe to the mission progress.
+    // We pass on device to receive_mission_progress because we need it in the callback.
+    device.mission().subscribe_progress(std::bind(&receive_mission_progress, _1, _2, &device));
 
-    if (result == Mission::Result::SUCCESS) {
-        _mission_state = MissionState::UPLOADING_DONE;
-    } else {
-        std::cerr << "Error: mission send result: " << unsigned(result) << std::endl;
-        _mission_state = MissionState::ERROR;
-    }
-}
-
-void receive_start_mission_result(Mission::Result result)
-{
-    EXPECT_EQ(result, Mission::Result::SUCCESS);
+    {
+        auto prom = std::make_shared<std::promise<void>>();
+        auto future_result = prom->get_future();
+        device.mission().start_mission_async(
+        [prom](Mission::Result result) {
+            EXPECT_EQ(result, Mission::Result::SUCCESS);
+            prom->set_value();
+        });
 
-    if (result == Mission::Result::SUCCESS) {
-        _mission_state = MissionState::STARTING_DONE;
-    } else {
-        std::cerr << "Error: mission start result: " << unsigned(result) << std::endl;
-        _mission_state = MissionState::ERROR;
+        future_result.wait();
+        future_result.get();
     }
-}
 
-void receive_pause_mission_result(Mission::Result result)
-{
-    EXPECT_EQ(result, Mission::Result::SUCCESS);
+    // Make sure we have taken off before checking for landed.
+    std::this_thread::sleep_for(std::chrono::seconds(5));
 
-    if (result == Mission::Result::SUCCESS) {
-        _mission_state = MissionState::PAUSE;
-    } else {
-        std::cerr << "Error: mission start result: " << unsigned(result) << std::endl;
-        _mission_state = MissionState::ERROR;
-    }
-}
-
-void receive_continue_mission_result(Mission::Result result)
-{
-    EXPECT_EQ(result, Mission::Result::SUCCESS);
+    // FIXME: this check does currently not work in SITL.
+    while (device.telemetry().in_air()) {
```

Comment:
You could try check on `disarmed()`. Worked on last example.

Reply:
Good idea, thanks.